class.c

```diff
@@ -1364,9 +1377,19 @@ class_descendants_recursive(VALUE klass, VALUE ary)
 VALUE
 rb_class_descendants(VALUE klass)
 {
-    VALUE ary = rb_ary_new();
-    rb_class_foreach_subclass(klass, class_descendants_recursive, ary);
-    return ary;
+    struct subclass_traverse_data data = { NULL, 0 };
+
+    // estimate the count of subclasses
+    rb_class_foreach_subclass(klass, class_descendants_recursive, (VALUE) &data);
+
+    // this allocation may cause GC which may reduce the subclasses
```

Comment:
I guess technically GC might also increase the number of subclasses, e.g., if a finalizer define new classes, no?

Reply:
I agree that could be an issue.  We could store the count of the first pass and either stop or raise an exception if we go beyond that count during the second pass to avoid writing into unallocated memory.

Reply:
Thanks. I think I have fixed the issue as @jeremyevans said. I'm a bit unsure if a new subclass can be inserted in the middle of the list instead of the last. If it is possible, `Class#descendants` may miss some subclasses, but I have no idea to fix the issue.